src/biokbase/narrative/tests/test_app_util.py

```diff
@@ -331,37 +334,175 @@ def test_transform_param_value_textsubdata(value, expected):
     assert transform_param_value(None, value, spec) == expected
 
 
-mock_obj_ref = "wjriehl:1490995018528/Sbicolor2"
-ref_cases = [
-    (None, None),
-    ("foo/bar", "foo/bar"),
-    ("1/2/3", mock_obj_ref),
-    ("Sbicolor2", mock_obj_ref),
+tf_types = [None, "ref", "resolved-ref", "unresolved-ref", "putative-ref", "upa"]
+all_types = []
+for t in tf_types:
+    all_types.append(t)
+    if t is not None:
+        all_types.append(f"list<{t}>")
+
+ws_name = "narrative_1695311690413"
+
+VALID_UPA_A = "69356/5/2"
+VALID_UPA_B = "69356/4/1"
+VALID_NAME_A = "Abiotrophia_defectiva_ATCC_49176"
+VALID_NAME_B = "_Nostoc_azollae__0708"
+VALID_REF_A = f"{user_name}:{ws_name}/Abiotrophia_defectiva_ATCC_49176"
+VALID_REF_B = f"{user_name}:{ws_name}/_Nostoc_azollae__0708"
+
+input_list = [
+    pytest.param(None, id="none"),
+    pytest.param(VALID_UPA_A, id="upa"),
+    pytest.param(VALID_NAME_A, id="name"),
+    pytest.param(VALID_REF_A, id="ws_name"),
+    pytest.param([VALID_UPA_A, VALID_UPA_B], id="upa_arr"),
+    pytest.param([VALID_NAME_A, VALID_NAME_B], id="name_arr"),
+    pytest.param(
+        [
+            VALID_REF_A,
+            VALID_REF_B,
+        ],
+        id="ws_name_arr",
+    ),
 ]
 
 
-@mock.patch("biokbase.narrative.app_util.clients.get", get_mock_client)
-@pytest.mark.parametrize("value,expected", ref_cases)
-def test_transform_param_value_simple_ref(value, expected, workspace_name):
-    workspace_name(workspace)
-    for tf_type in ["ref", "unresolved-ref"]:
-        tf_value = transform_param_value(tf_type, value, None)
-        assert tf_value == expected
+@pytest.mark.parametrize("tf_type", all_types)
+@pytest.mark.parametrize("inp", input_list)
+def test_transform_param_value_all_types_with_all_inputs(
+    inp, tf_type, workspace_name, monkeypatch, request
+):
+    workspace_name(f"{user_name}:{ws_name}")
 
+    def get_workspace(_):
+        # ensure that there's a token for the workspace client
+        return Workspace(URLS.workspace, token=user_token)
 
-mock_upa = "18836/5/1"
-upa_cases = [
-    ("some_name", mock_upa),
-    (["multiple", "names"], [mock_upa, mock_upa]),
-    ("already/ref", mock_upa),
+    monkeypatch.setattr(clients, "get", get_workspace)
```

Comment:
Unfortunately when creating the WS client in app_utils, the token is not added, so requests for object info get sent without any auth info. This patches the creation of the client to ensure there is a token present.